File: KthLargestElement.py
# TC: O(k)
# SC: O(h)
# Inorder traversal
# Iterative
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        if not root: return None
        stack = []
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            k -= 1
            if k == 0:
                return root.val
            root.right
                
# Recursive
class Solution:
    result = None
    count = 0
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        if not root: return None
        self.helper(root, k)
        return self.result
    
    def helper(self, root, k):
        # base
        if root != None:
            print(root.val)
        if root == None:
            return
        # logic
        self.helper(root.left, k)
        self.count += 1
        if k == self.count:
            self.result = root.val
            return
        # small optimization
        if self.result == None:
            self.helper(root.right, k)

File: KthLargestElement.java
//TC : O(log(n))
//SC : O(1)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
/*class Solution {
    PriorityQueue<TreeNode> pq;
    public int kthSmallest(TreeNode root, int k) {
        pq = new PriorityQueue<>((a,b)->(b.val-a.val));
        helper(root,k);
        return pq.poll().val;
    }
    
    private void helper(TreeNode root, int k){
        if(root==null) return;
        pq.add(root);
        if(pq.size() > k){
            pq.poll();
        }
        helper(root.left,k);
        helper(root.right,k);
    }
}*/
class Solution {
    int cnt;
    int result;
    boolean flag;
    public int kthSmallest(TreeNode root, int k) {
       cnt =0;
       flag=false;
        inorder(root,k);
        return result;
    }
    
    private void inorder(TreeNode root, int k){
        if(root == null) return;
        if(!flag)inorder(root.left,k);
        cnt++;
        if(cnt==k){
            flag=true;
            result = root.val;
            return;
        }
        if(!flag)inorder(root.right,k);
    }
}
